### PR TITLE
Version 0.2.0 proposal 

### LINKED ISSUE #4: Null response without an error
**Describe the bug**
So when you trying to load up a schematic via Scaffolding.fromFile or via the FileInputStream the response is always null. There is no exception too which would be thrown. BTW my entire completeable future is beeing closed but i have no idea what why and whats caused it...

**To Reproduce**
1. Load up a schematic
2. look via if(schematic != null) if its null
3. Use it in a CompleteableFuture
4. See null response with no error (Schematics are tested and worked fine)

**Expected behavior**
It should throw an error and should load when the schematics are right, load them up.

**Screenshots**
1. 
image
2. (Debug) 
image
(- Please ask for a full code snippet)

**Desktop (please complete the following information):**
 - Latest Scaffolding version
 - Latest Minestom verison (Not 1.18 branch)
 - Java 17

### LINKED ISSUE #8: Cannot place schematic
**Describe the bug**
When placing a schematic it produces:
```
java.lang.ArrayIndexOutOfBoundsException: Index 1048576 out of bounds for length 24
	at java.base/java.util.ImmutableCollections$ListN.get(ImmutableCollections.java:680)
	at net.minestom.server.instance.DynamicChunk.getSection(DynamicChunk.java:109)
	at net.minestom.server.instance.Chunk.getSectionAt(Chunk.java:95)
	at net.minestom.server.instance.DynamicChunk.setBlock(DynamicChunk.java:72)
	at net.minestom.server.instance.batch.ChunkBatch.apply(ChunkBatch.java:216)
	at net.minestom.server.instance.batch.ChunkBatch.singleThreadFlush(ChunkBatch.java:188)
	at net.minestom.server.instance.batch.ChunkBatch.lambda$apply$0(ChunkBatch.java:156)
```

**To Reproduce**
```
Schematic schem = Scaffolding.fromFile(...);
schem.build(..., Pos.ZERO);
```
(unpack for schematic, WE is ok with it, and it seems to have everything expected)
[lightouse.schem.zip](https://github.com/CrystalGamesMC/Scaffolding/files/8311313/lightouse.schem.zip)

**Expected behavior**
should build

**Screenshots**

**Desktop (please complete the following information):**
ah yes very good

**Smartphone (please complete the following information):**
+1 for template

**Additional context**

### LINKED ISSUE #10: MCEdit schematics use the incorrect block pallet
**Describe the bug**
MCEdit schematics are loaded with a different block pallet when compared to how WorldEdit handles the same schematic. Currently, state Ids can only go from 0 to 255 if addBlocks is not used. This is an issue since Minestom block state Ids can go up to ~15000. To fix the issue, blockId and blockData should be mapped together to get the true stateId, such as how [WorldEdit does](https://github.com/EngineHub/WorldEdit/blob/master/worldedit-core/src/main/resources/com/sk89q/worldedit/world/registry/legacy.json).

I have a have a solution that just needs to be cleaned up when I have a bit of free time. I just wanted to create an issue so I can resolve it ;)

### PR DESCRIPTION
**Pull request by DamianKocher**
This PR fixes many issues but probably creates many more. I believe these changes are necessary to reach a point where this library can be used in a production capacity. 

**Fixes**
Fixes #4 - Scaffolding methods now cannot return null
Fixes #8 - Schematics now cannot be placed in positions which would put them outside of world bounds
Fixes #10 - MCEdit schematics now use correct block palette 

**Improvements**
- More documentation
- Distribution via JitPack is supported
- Use CompletableFutures wherever possible
- Rewrite schematic loading to make adding new schematic formats easier
- WIP wiki site
- WIP WorldEdit like editor server
- Many other small changes

### CODE REVIEW

**Comment by LooFifteen** [comment]
you madman

**Comment by DamianKocher** [comment]
> you madman

yeah i might've gotten a bit carried away lmao

File: .gitignore [hunk 1]
```diff
@@ -36,14 +36,14 @@
...
-*.iml
-*.ipr
+# .idea/artifacts
+# .idea/compiler.xml
+# .idea/jarRepositories.xml
+# .idea/modules.xml
+# .idea/*.iml
+# .idea/modules
+# *.iml
+# *.ipr
```
Comment (LooFifteen) [inline .gitignore]: This stuff should be uncommented.

File: .gitignore [hunk 2]
```diff
@@ -118,32 +114,39 @@ modules.xml
...
 extensions/
 
 ### Scaffolding ###
 *.schematic
 *.schem
 
-# End of https://www.toptal.com/developers/gitignore/api/maven,intellij+all,java
\ No newline at end of file
+### Let's just ignore the whole .idea folder for now###
+.idea/
```
Comment (LooFifteen) [inline .gitignore]: hmmmm

File: scaffolding-core/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+description = "Schematic library for Minestom"
+java.sourceCompatibility = JavaVersion.VERSION_17
+
+var minestomVersion = "bfa2dbd3f7"
+
+dependencies {
+    compileOnly("com.github.Minestom:Minestom:${minestomVersion}")
+    testImplementation("com.github.Minestom:Minestom:${minestomVersion}")
+
+    testImplementation("me.alexpanov:free-port-finder:1.1.1")
```
Comment (LooFifteen) [inline scaffolding-core/build.gradle.kts]: Should probably just use port 25565 for convenience sake.

File: scaffolding-core/src/main/java/net/crystalgames/scaffolding/Scaffolding.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import org.jglrxavpok.hephaistos.nbt.NBTException;
+import org.jglrxavpok.hephaistos.nbt.NBTReader;
+
+import java.io.*;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A static utility class primarily used to parse schematics.
+ */
+@SuppressWarnings("unused")
```
Comment (LooFifteen) [inline scaffolding-core/src/main/java/net/crystalgames/scaffolding/Scaffolding.java:19]: Warnings probably shouldn't be suppressed. Unused is a valid thing to supress but I prefer keeping warnings in.

File: scaffolding-core/src/main/java/net/crystalgames/scaffolding/region/Region.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+
+/**
+ * Represents a rectangle 3 dimensional region of blocks withing an {@link Instance}.
+ */
+@SuppressWarnings("unused")
```
Comment (LooFifteen) [inline scaffolding-core/src/main/java/net/crystalgames/scaffolding/region/Region.java]: Same suppressing thing here.

File: scaffolding-editor/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+description = "Scaffolding Editor"
+java.sourceCompatibility = JavaVersion.VERSION_17
+
+var minestomVersion = "f774cc3b0f"
+
+dependencies {
+    implementation(project(":scaffolding-core"))
+
+    implementation("me.alexpanov:free-port-finder:1.1.1")
```
Comment (LooFifteen) [inline scaffolding-editor/build.gradle.kts]: Should also probably use 25565

**Review by LooFifteen — CHANGES_REQUESTED** [review]
Formatting is a bit odd with the lack of spaces at the end of files, other than that and a few other comments, looking really good and quite interesting.

Was actually planning to migrate to Gradle, thank goodness someone did it for me.

File: scaffolding-core/src/main/java/net/crystalgames/scaffolding/Scaffolding.java (hunk 4 above)
Comment (DamianKocher) [reply scaffolding-core/src/main/java/net/crystalgames/scaffolding/Scaffolding.java:19]: You're right. If I ever get around to adding proper unit tests suppressing unused warnings won't be necessary. I'll remove them.

**Comment by LooFifteen** [comment]
Should the server have Moulberry's WorldEdit installed by default instead of recreating a selection tool? It should allow the end user to have more editing capability.

**Comment by DamianKocher** [comment]
I thought about that. I'm a bit torn on the issue. One one hand, leveraging the 10+ years of development that WE has is a smart move. On the other, I think with enough work it would be possible to surpass the capabilities of what WE has to offer.

I personally would like to go down the second route, but if you want me to cut the editor in favor of using WE I understand.

**Comment by LooFifteen** [comment]
If you have the time and dedication to do that, go ahead. I'm unable to spend that time unfortunately.

**Comment by joshuasing** [comment]
Hey @DamianKocher, thank you for contributing, I am able to work on this in my free time so I will be merge this into a development branch to continue.

**Pull request closed without merging**